gitpod-io/gitpod
# hook scrubber as logrus formatter

## Description


We reconsidered approach and decided to redact all nested properties if they don't implement `Scrubbable` interface.

<details>
  <summary>Previous attempt</summary> I tried first to scrub before formatting and wasted too much time tuning scrubbing itself (although did fix some cases). 
  But it turned out that scrubbing mutates data as for now, so we have to make the deep copy otherwise it breaks for 
  instance startup of image build and public api. So I went for JSON formatting first and then scrubbing of JSON. It implies 
  though additional unmarshalling/marshalling. It makes logging 2-3x times slower. I am not sure we should invest more 
  time in tuning since benchmarking still shows that on average for complex data structures it takes 10000 ns.
</details>

I also explored using logurs hooks, but they don't have impact on the control flow, see [[1](https://github.com/sirupsen/logrus/blob/352781de903c9dc639752a3ac08148132746e180/entry.go#L242)]

## Related Issue(s)


fix SID-285

## How to test


- Start a workspace.
- Inspect different go component logs for some noisy errors when scrubbing prevents logging.
- Check public api logs for instance for redact.

## Documentation


#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - ak-logrus-scrub</li>
	<li><b>🔗 URL</b> - <a href="https://ak-logrus-scrub.preview.gitpod-dev.com/workspaces" target="_blank">ak-logrus-scrub.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

      Run the build with werft instead of GHA
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

</details>

<details>
<summary>Installer</summary>

  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

      If enabled this will build `install/preview`
      If enabled this will create the environment on GCE infra
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>
 
/hold
